### ISSUE TITLE
bug in qx.ui.table.cellrenderer.Number

### ISSUE THREAD

**Issue by dvoral** [qooxdoo/qooxdoo#9340]
If the cell is empty (null) and is being edited, then NaN displays. 

From Derrell Lipman:
if (cellInfo.value || cellInfo.value == 0) {
    should be
if (cellInfo.value || cellInfo.value === 0) {
Note the triple '=' so that it catches the number 0 but not something that's falsy like null.

We should also change the else condition where it says cellInfo.value == 0.

**Comment by johnspackman** [comment]
Hi @dvoral, thanks for the bug report - could you try implementing a Pull Request with this change?  That's the fastest way to get a fix integrated into the core

**Comment by derrell** [comment]
Sigh. There is no good solution to this problem.

On the one hand, we can change the `format` method of `qx.util.format.NumberFormat` so that an empty string is not NaN, but that is not backward compatible with apps that expect an empty string to return NaN.

We could also change qx.ui.table.cellrenderer.Number so that it does what its name implies and ensures that every field contains a number. This would mean that if the data is null or an empty string, we put a 0 in the cell. That way, the cell editor (which was the basis of the OP's problem for which this issue was created) would always get a valid initial value, and not display NaN. That is also not backbackward compatible, because programs may well expect the ability to have empty fields in a Number column.

I'm open to ideas.

Derrell

**Comment by zaucker** [comment]
A number column in a table used for data entry MUST allow empty cells.

How about a separat cellrenderer class? That would not break anything.

Cheers,
Fritz

**Comment by level420** [comment]
@dvoral did you find a solution for this bug? Please share you're solution/thoughts. Thank you.

**Comment by dvoral** [comment]
Yes I did. I wrote on the forum what I did. On June 4 I wrote:
"I managed to fix the NaN while editing problem. I wrote an override of the
setValidationFunction of the qx.ui.table.celleditor.TextField."

Nobody asked until now what I did.
My solution:
var intEditor = new qx.ui.table.celleditor.TextField();
intEditor.setValidationFunction (
               function( value ) {
                if (value == "") {
                value = null;
                }
                if (value != null) {
                value = parseFloat(value);
                }
                return value;
               });


Hope this is helpful.

**Comment by level420** [comment]
@dvoral would you please be so kind to create a pull request for `qx.ui.table.celleditor.TextField`. that would be great! thank you!

**Comment by dvoral** [comment]
Can you please send me a link where I do that?
And instructions if it's not clear for a first-time user. :)

**Comment by derrell** [comment]
I'd suggest adding this as a static function `numberOrWhitespace` (or some such thing) in the cell renderer, so that it would be easy to just do something like
```
intEditor.setValidationFunction(intEditor.constructor.numberOrWhitespace)
```

**Comment by level420** [comment]
@dvoral there are several tutorials out there but here a short version which works completely in the browser:

* Create a fork of qooxdoo master (the button called `fork` on the top right of this page). This creates a fork of qooxdoo master under your account called dvoral/qooxdoo
* go to repo dvoral/qooxdoo and navigate in the source tree to `framework/source/class/qx/ui/table/celleditor/TextField.js`
* on the top right of the source code view there is a pencil icon. click it to open the file in edit mode
* make your changes and add comments if needed 
* save the changes (there is a button below of the editor). you will be asked if you want to save it in master or in some branch. choose the branch and name it e.g. with the issue # in the name, something like dvorak-fix-9340
* after sucessfully saving navigate to the master branch of your qooxdoo fork (https://github.com/level420/qooxdoo)
* you'll note a green "create pull request" button which, if clicked, offers you to create a PR which is against qooxdoo/qooxdo master from dvorak/qooxoo/<howyounamedthebranch>

That's it.

Try it! You could do no harm to the original repo. We won't let you :-)

**Comment by dvoral** [comment]
Thanks for your detailed instructions. I've made the pull request.

**Comment by johnspackman** [comment]
Hi,

Thank you for your email - please note that I am on annual leave, returning on June 21st  
 If you have an emergency support issue, please describe your issue in detail in email and then send a text message to the mobile number on your invoice.

Regards
John Spackman

**Comment by level420** [comment]
@dvoral you create the PR from dvoral/qooxdoo:dvoral-fix-9340 against your own forks master dvoral/qooxdoo:master

Please navigate again to https://github.com/dvoral/qooxdoo/pulls to the PR list and then click the green button "New pull request".

There will appear four buttons for selection on the "Comparing changes" page.

Please select there from the rightmost button (Text beginning with "compare:") the dovarl-fix-9340 branch. This way you compare your branch against qooxdoo/qooxdoo:master.

You will then be offered again the "Create pull request" green button which allows you to publish your PR in https://github.com/qooxdoo/qooxdoo/pulls

**Comment by dvoral** [comment]
Hope I did it correctly this time.

**Comment by level420** [comment]
@dvoral if you need to add changes to the PR or change your current PR, you simply navigate to the corresponding branch https://github.com/dvoral/qooxdoo/tree/dvoral-fix-9340 and edit what you want and save to the same branch (here dvoral-fix-9340).

As soon as you've done that, the changes appear in the PR https://github.com/qooxdoo/qooxdoo/pull/9348

**Comment by dvoral** [comment]
I changed the place of the check on null string as recommended.

**Comment by level420** [comment]
Fixed via #9348
